### PR TITLE
WIP: config rewrite

### PR DESCRIPTION
**Pull request by Rick-Anderson**

### CODE REVIEW

**Comment by Rick-Anderson** [comment]
@HaoK  are you the right person to review this? Do you want me to send you a word doc? cc @tdykstra

**Comment by Rick-Anderson** [comment]
Word version at https://www.dropbox.com/s/ujk0jk5i27kfywo/Configuration.docx?dl=0 if that's easier.  @HaoK  can you suggestion someone to do a quick review?

**Comment by HaoK** [comment]
@divega @glennc should also review, I will take a look as well

**Comment by Rick-Anderson** [comment]
@HaoK  @divega @glennc can you give me an ETA for the review?

Word version at https://www.dropbox.com/s/ujk0jk5i27kfywo/Configuration.docx?dl=0 if that's easier.  I could also send you the HTML version.

**Comment by HaoK** [comment]
You should update the startup code in general to use the new pattern:

```
        public Startup(IHostingEnvironment env)
        {
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
```

instead of Directory.GetCurrentPath() to be consistent with the templates

**Comment by HaoK** [comment]
I did a first pass, and commented some in the word doc via dropbox, you should get @divega to do a final review as well

File: aspnet/fundamentals/configuration.rst [hunk 1]
```diff
@@ -2,169 +2,252 @@
...
 
-The ``Index`` view displays the configured options:
+.. literalinclude:: configuration/sample/src/WebConfigBind/Controllers/HomeController.cs
+  :language: none
 
-.. image:: configuration/_static/index-view.png
+.. _custom-config-providers:
 
-You configure options using the :dn:method:`~Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.Configure\<TOptions>` extension method. You can configure options using a delegate or by binding your options to configuration:
+Note that Dependency Injection (DI) is not setup until after ``ConfigureServices`` is invoked and the configuration system is not DI aware.
```
Comment (tdykstra) [inline aspnet/fundamentals/configuration.rst]: Not clear why this note is here as the issue is not directly relevant to this particular example.  Also the sentence structure is ambiguous, and it's not clear what "not DI aware" means.

File: aspnet/fundamentals/configuration.rst [hunk 2]
```diff
@@ -2,169 +2,252 @@
...
-Example: Entity Framework Settings
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Entity Framework custom provider
+---------------------------------
 
-You may wish to store some of your application's settings in a database, and access them using Entity Framework Core (EF). There are many ways in which you could choose to store such values, ranging from a simple table with a column for the setting name and another column for the setting value, to having separate columns for each setting value. In this example, we're going to create a simple configuration provider that reads name-value pairs from a database using EF.
+In this section we'll create a simple configuration provider that reads name-value pairs from a database using EF.
 
-To start off we'll define a simple ``ConfigurationValue`` entity for storing configuration values in the database:
+Define a `ConfigurationValue`` entity for storing configuration values in the database:
```
Comment (tdykstra) [inline aspnet/fundamentals/configuration.rst]: extra backtick

**Comment by Rick-Anderson** [comment]
@HaoK  wrote:

You should update the startup code in general to use the new pattern:

```
        public Startup(IHostingEnvironment env)
        {
            var builder = new ConfigurationBuilder()
                .SetBasePath(env.ContentRootPath)
```

instead of Directory.GetCurrentPath() to be consistent with the templates

I'm guessing you meant `Directory.GetCurrentDirectory()`, not `Directory.GetCurrentPath()`
Do you remember where you saw this? I only have two Web projects, and they both use the pattern in the RTM templates.

File: aspnet/fundamentals/configuration.rst [hunk 3]
```diff
@@ -2,169 +2,252 @@
...
+.. _options-example:
 
-.. warning:: You should never store passwords or other sensitive data in configuration provider code or in plain text configuration files. You also shouldn't use production secrets in your development or test environments. Instead, such secrets should be specified outside the project tree, so they cannot be accidentally committed into the configuration provider repository. Learn more about :doc:`environments` and managing :doc:`/security/app-secrets`.
+In the code below, ``ConfigurationBuilder`` is initialized in the ``Startup`` class. In the ``ConfigureServices`` method, an ``IConfigurationRoot`` interface is passed to the to :dn:method:`~Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.Configure\<TOptions>`, which adds ``ConfigurationBuilder`` to the dependency injection container:
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/configuration.rst]: Per the style guide I'm removing the description and moving all code explanations to code comments.

>  Why do we use IConfigurationRoot here but we used IConfiguration in the first snippet in the article?

Nice catch. It should be IConfigurationRoot  everywhere, we want the root of the configuration.  IConfiguration is the lower level interface that is specialized by `IConfigurationRoot` and `IConfigurationSection`

File: aspnet/fundamentals/configuration.rst [hunk 4]
```diff
@@ -2,169 +2,252 @@
...
+There is no limit to the number of ``IConfigureOptions<TOptions>`` services you can add. Each configuration service comes in a NuGet package. They are all applied in order they are registered. Each call to :dn:method:`~Microsoft.Extensions.DependencyInjection.OptionsServiceCollectionExtensions.Configure\<TOptions>` adds an :dn:iface:`~Microsoft.Extensions.Options.IConfigureOptions\<TOptions>` service to the service container. In the example above, the values of ``Option1`` and ``Option2`` are both specified in `appsettings.json`, but the value of ``Option1`` is overridden by the configured delegate in the highlighted code above. When more than one configuration service is enabled, the last configuration source specified “wins”. With the code above, the ``HomeController.Index`` method returns ``option1 = value1_from_action, option2 = 2``.
 
-A simple ``MyOptions`` class is shown here:
+.. note:: Configuration keys are case insensitive.
```
Comment (Rick-Anderson) [inline aspnet/fundamentals/configuration.rst]: Moved it up, removed note. Remove 2 notes, 1 tip and 1 warning all to a bullet list.

File: aspnet/fundamentals/configuration.rst (hunk 1 above)
Comment (Rick-Anderson) [inline aspnet/fundamentals/configuration.rst]: moved to bottom.  This is a copy/paste from Dan.Roth, folks that need to know will figure it out.

**Comment by Rick-Anderson** [comment]
@HaoK per your DB feedback (I fixed all  your suggestions):

- Spit out `GetValue `in it's own section
- Explained `GetValue` is basically just syntax sugar for `GetSection(<key>).TypeConverter.Convert<T>` and it's not doing any binding
- Added ctor to models and explained that's normally how you set defaults.

@divega @glennc can you review? I can send you a word file if you like.  I'll need a shipit

**Comment by Rick-Anderson** [comment]
@HaoK  I added:

The built in configuration providers are not refreshed when the configuration data changes. If your configuration data changes, you'll need to restart your app to get the new data.

Now how do I explain `reloadOnChange`?
```
public static IConfigurationBuilder AddJsonFile(this IConfigurationBuilder builder,
                                           string path, bool optional, bool reloadOnChange)

```

**Comment by Rick-Anderson** [comment]
see https://github.com/aspnet/Configuration/issues/532

**Comment by Rick-Anderson** [comment]
moved to #2120  -

**Pull request closed without merging**